[LoopCacheAnalysis] Fix loop cache cost to always round the cost up to the nearest integer number

RouzbehPaktinat opened the pull request:
Currently loop cache analysis uses following formula to evaluate cost of an RefGroup for a consecutive memory access: 

`RefCost=(TripCount*Stride)/CLS`

This cost evaluates to zero when `TripCount*Stride` is smaller than cache-line-size. This results in wrong cost value for a loop and misleads loopInterchange decisions as shown in [this case](https://llvm.godbolt.org/z/jTz1vn4hn).

This patch fixes the problem by rounding the cost to 1 once this problem happens.

CongzheUalberta commented:
There are test case failures. Can you please check?

nikic commented on llvm/lib/Analysis/LoopCacheAnalysis.cpp:
```diff
@@ -299,8 +299,15 @@ CacheCostTy IndexedReference::computeRefCost(const Loop &L,
+    // be used. It does not make sense to output the result that zero cache line
+    // is used
+    RefCost =
+        IsZero ? SE.getSCEV(One) : SE.getUDivExpr(Numerator, CacheLineSize);
```
I am not familiar with this code, but possibly what you are looking for is actually getUDivCeilSCEV? That is round the division up instead of down.

RouzbehPaktinat replied:
Hi @nikic 
Thank you for your reviews. I made both changes. I used `getUDivCeilSCEV `only when refCost evaluates to zero. If I completely replace `getUDivExpr` with `getUDivCeilSCEV` many tests fail because it's now rounding all cost values not just zeros which is not purpose of this PR.

nikic replied:
Using getUDivCeil after checking for zero doesn't really make sense -- at that point, just directly setting it to one would be better.

That many tests require changes when using getUDivCeil is expected -- the question is whether those changes are incorrect or not. Without being familiar with the code, I would expect that the round-up division is what is correct in this context.

CongzheUalberta reviewed (commented):
There are test case failures. Can you please check?

RouzbehPaktinat commented:
> There are test case failures. Can you please check?

It's fixed now.

CongzheUalberta reviewed (approved):
LGTM. Please check if other reviewers are okay with your updates, before you merge the patch.

RouzbehPaktinat commented:
Gently ping @nikic

nikic reviewed (commented):
This looks reasonable to me, but it's probably best if @CongzheUalberta takes another look.

CongzheUalberta commented on llvm/lib/Analysis/LoopCacheAnalysis.cpp:
```diff
@@ -299,7 +299,10 @@ CacheCostTy IndexedReference::computeRefCost(const Loop &L,
-    RefCost = SE.getUDivExpr(Numerator, CacheLineSize);
+    // Round the costs up to the nearest whole number i.e. when cost is
+    // calculated 12.5, it makes more sense to say 13 cache lines are used
+    // rather than 12 cache lines.
```
// Round the cost up to the nearest integer number, e.g., when cost is
// calculated to be a number less than one, it makes more sense to say 
// one cache line is used rather than zero cache line is used.

CongzheUalberta commented:
Please squash your commits, and also update your commit messages to reflect your most recent changes (you are now rounding up the cost not only from 0 to 1). There's typo in your commit message as well (e.g., "stride"), please double check.

RouzbehPaktinat committed:
[LoopCacheAnalysis] Fix loop cache cost to always round the cost up to the nearest integer number.

Currently loop cache analysis uses following formula to evaluate cost of an RefGroup for a consecutive memory access: "RefCost=(TripCount*Stride)/CLS". When the cost is fractional, it's always rounded to the smaller integer number but this is problematic specially when the cost is calculated to be a number less than one which will be rounded to zero. But it makes more sense to say one cache line is used in this case rather than zero cache lines.

This patch fixes the problem by always rounding the cost up to the nearest larger integer number.

RouzbehPaktinat commented:
Thank you for your reviews @nikic @CongzheUalberta 
I addressed new comments.

CongzheUalberta reviewed (approved):
Thanks for the updates, LGTM.

chenzheng1030 reviewed (commented):
Would you mind wait for some days for this patch? @RouzbehPaktinat 
Most changes happen inside PowerPC target. I will add some PowerPC reviewers and ping them for review. Thanks.

chenzheng1030 commented:
> I didn't see anything in the PPC LIT changes that I think is alarming. I think this is okay.

Thanks for checking, Roland.

chenzheng1030 reviewed (approved):
Thanks very much for your patience.

Narutoworld commented:
Thank you all for the reviews. We'll merge it when Rouzbeh is back from vacation.

RouzbehPaktinat commented:
Thank you everyone for your reviews and comments.
